### PR TITLE
Add metadata filter handling for builtin source storages

### LINKED ISSUE #423: use metadata filter for chats
### Feature description

See #422 for work that has to happen before this. See #256 for a discussion.

- `SourceStorage.retrieve` needs to take a `MetadataFilter` object as first input argument rather than a list of documents
- The builtin source storages need to be refactored in a way to only have a single index. The metadata of each document should be stored alongside the document in the DB. During retrieval, the metadata filter needs to be translated into the dialect of the DB
- The first input argument of `Rag().chat()` needs to be renamed to `input` and should allow a list of `Document`s as well as a `MetadataFilter`:
  - Passing a list of documents should retain the current behavior by storing them in the source storage and creating a metadata filter to select only these documents in the retrieval stage.
  - Passing a metadata filter should skip the storing and just set the prepared flag on the chat
  The cases of passing both types or none as explored in https://github.com/Quansight/ragna/issues/256#issuecomment-1954057313 will be saved for later. For now only a single input is fine
- For the REST API, instead of overloading one input parameter as done for the Python API, we should have two mutually exclusive `documents` and `metadata_filter` inputs.

### Value and/or benefit

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by pmeier**
Closes #423.

### CODE REVIEW

**Comment by pmeier** [comment]
Chroma and LanceDB are defunct for now. Here is a demo using `ragna.source_storages.RagnaDemoSourceStorage`:

<details><summary>full script</summary>
<p>

```python
from __future__ import annotations

import uuid
from typing import Any, Iterator

from ragna import Rag, source_storages, core
from ragna.core import DocumentUploadParameters, Source
from ragna.deploy import Config


class DummyDocument(core.Document):
    async def get_upload_info(
        cls, *, config: Config, user: str, id: uuid.UUID, name: str
    ) -> tuple[dict[str, Any], DocumentUploadParameters]:
        raise NotImplementedError

    @classmethod
    def new(cls, name: str, **metadata) -> DummyDocument:
        return cls(
            name=name,
            metadata=metadata,
            handler=core.PlainTextDocumentHandler(),
        )

    def is_readable(self) -> bool:
        return True

    def read(self) -> bytes:
        return b""


class DummyAssistant(core.Assistant):
    def answer(self, prompt: str, sources: list[Source]) -> Iterator[str]:
        yield "\n".join(f"- {source.document_name}" for source in sources)


async def main(metadatas, metadata_filters):
    documents = [
        DummyDocument.new(f"document{idx}.txt", idx=idx, **metadata)
        for idx, metadata in enumerate(metadatas)
    ]
    for document in documents:
        print(f"- {document.name}: {document.metadata}")

    source_storage = source_storages.RagnaDemoSourceStorage()
    source_storage.store(documents)

    for metadata_filter in metadata_filters:
        print("-" * 80)
        print(metadata_filter)
        print()

        chat = Rag().chat(
            metadata_filter, source_storage=source_storage, assistant=DummyAssistant
        )
        answer = await chat.answer("?")
        print(answer)


metadatas = [
    {
        "priority": "low",
        "department": "legal",
    },
    {
        "priority": "medium",
        "department": "legal",
    },
    {
        "priority": "low",
        "department": "marketing",
    },
    {
        "priority": "high",
        "department": "marketing",
    },
    {
        "priority": "medium",
        "department": "marketing",
    },
]

metadata_filters = [
    core.MetadataFilter.eq("document_name", "document2.txt"),
    core.MetadataFilter.ge("idx", 3),
    core.MetadataFilter.eq("department", "legal"),
    core.MetadataFilter.in_("priority", ["medium", "high"]),
    core.MetadataFilter.and_(
        [
            core.MetadataFilter.eq("priority", "low"),
            core.MetadataFilter.eq("department", "legal"),
        ]
    ),
    core.MetadataFilter.and_(
        [
            core.MetadataFilter.or_(
                [
                    core.MetadataFilter.eq("priority", "medium"),
                    core.MetadataFilter.eq("department", "marketing"),
                ]
            ),
            core.MetadataFilter.lt("idx", 4),
        ]
    ),
]

import asyncio

asyncio.run(main(metadatas, metadata_filters))
```

</p>
</details> 

```python
[...]
metadata_filters = [
    core.MetadataFilter.eq("document_name", "document2.txt"),
    core.MetadataFilter.ge("idx", 3),
    core.MetadataFilter.eq("department", "legal"),
    core.MetadataFilter.in_("priority", ["medium", "high"]),
    core.MetadataFilter.and_(
        [
            core.MetadataFilter.eq("priority", "low"),
            core.MetadataFilter.eq("department", "legal"),
        ]
    ),
    core.MetadataFilter.and_(
        [
            core.MetadataFilter.or_(
                [
                    core.MetadataFilter.eq("priority", "medium"),
                    core.MetadataFilter.eq("department", "marketing"),
                ]
            ),
            core.MetadataFilter.lt("idx", 4),
        ]
    ),
]
[...]
```

```
- document0.txt: {'idx': 0, 'priority': 'low', 'department': 'legal'}
- document1.txt: {'idx': 1, 'priority': 'medium', 'department': 'legal'}
- document2.txt: {'idx': 2, 'priority': 'low', 'department': 'marketing'}
- document3.txt: {'idx': 3, 'priority': 'high', 'department': 'marketing'}
- document4.txt: {'idx': 4, 'priority': 'medium', 'department': 'marketing'}
--------------------------------------------------------------------------------
EQ('document_name', 'document2.txt')

- document2.txt
--------------------------------------------------------------------------------
GE('idx', 3)

- document3.txt
- document4.txt
--------------------------------------------------------------------------------
EQ('department', 'legal')

- document0.txt
- document1.txt
--------------------------------------------------------------------------------
IN('priority', ['medium', 'high'])

- document1.txt
- document3.txt
- document4.txt
--------------------------------------------------------------------------------
AND(
  EQ('priority', 'low'),
  EQ('department', 'legal'),
)

- document0.txt
--------------------------------------------------------------------------------
AND(
  OR(
    EQ('priority', 'medium'),
    EQ('department', 'marketing'),
  ),
  LT('idx', 4),
)

- document1.txt
- document2.txt
- document3.txt
```

File: ragna/core/_rag.py [hunk 1]
```diff
@@ -232,9 +235,14 @@ async def answer(self, prompt: str, *, stream: bool = False) -> Message:
 
         return answer
 
-    def _parse_documents(self, documents: Iterable[Any]) -> list[Document]:
-        documents_ = []
-        for document in documents:
+    def _parse_input(
```
Comment (pmeier) [inline ragna/core/_rag.py:238]: @nenb @blakerosenthal Loosely leaning on https://github.com/Quansight/ragna/issues/256#issuecomment-1954057313, `input` is overloaded with 3 values:

1. If its explicitly `None`, i.e. no default value and actually calling `Chat(input=None)`, we don't want any metadata filtering and thus also `metadata_filter=None`. When we pass this down to the source storage, filtering should be disabled and thus the whole index should be used. _This is not implemented yet in this PR_
2. If its a `MetadataFilter`, we use it.
3. If its anything else, we assume it is list of documents or paths to create documents from. For a better UX, we might also allow a single paths or document, but this is not critical.

For 1. and 2. we assume the chat is already prepared and only require preparation for 3.

Comment (blakerosenthal) [reply ragna/core/_rag.py:238]: @pmeier @nenb What remains on the source storage work in particular? It looks like the chat object performs the logic of constructing a `MetadataFilter` object to pass along to source storage, and the source storage needs to handle the `metadata_filter=None` case and return the entire index. Is there anything else?

Comment (pmeier) [reply ragna/core/_rag.py:238]: Not sure if that is implied, but we need handling of the `MetadataFilter` in `Chroma` and `LanceDb`. But I think these are the open tasks right now.

Comment (nenb) [reply ragna/core/_rag.py:238]: @pmeier @blakerosenthal I had some uncertainty in the task as well.

I've opened #460. We can discuss what needs to be done there. And @blakerosenthal you can push the `lancedb` changes to the same branch.

**Comment by pmeier** [comment]
Merging this and use #460 and #461 as follow-ups.

**Pull request merged**